src/core/server/saved_objects/saved_objects_service.ts

```diff
@@ -0,0 +1,116 @@
+import { Logger } from '..';
+
+/**
+ * @public
+ */
+// eslint-disable-next-line @typescript-eslint/no-empty-interface
+export interface SavedObjectsServiceSetup {}
+
+/**
+ * @public
+ */
+export interface SavedObjectsServiceStart {
+  migrator: Pick<KibanaMigrator, keyof KibanaMigrator>;
+}
+
+/** @internal */
+export interface SavedObjectsSetupDeps {
+  legacy: LegacyServiceSetup;
+  elasticsearch: ElasticsearchServiceSetup;
+}
+
+/** @internal */
+// eslint-disable-next-line @typescript-eslint/no-empty-interface
+export interface SavedObjectsStartDeps {}
+
+export class SavedObjectsService
+  implements CoreService<SavedObjectsServiceSetup, SavedObjectsServiceStart> {
+  private migrator: KibanaMigrator | undefined;
+  logger: Logger;
+
+  constructor(private readonly coreContext: CoreContext) {
+    this.logger = coreContext.logger.get('savedobjects-service');
+  }
+
+  public async setup(coreSetup: SavedObjectsSetupDeps) {
+    this.logger.debug('Setting up SavedObjects service');
+
+    const {
+      savedObjectSchemas,
+      savedObjectMappings,
+      savedObjectMigrations,
+      savedObjectValidations,
+    } = await coreSetup.legacy.uiExports;
+
+    const adminClient = await coreSetup.elasticsearch.adminClient$.pipe(first()).toPromise();
+
+    const kibanaConfig = await this.coreContext.configService
+      .atPath<KibanaConfig>('kibana')
+      .pipe(first())
+      .toPromise();
+
+    this.migrator = new KibanaMigrator({
+      savedObjectSchemas,
+      savedObjectMappings,
+      savedObjectMigrations,
+      savedObjectValidations,
+      logger: this.coreContext.logger.get('migrations'),
+      kibanaVersion: this.coreContext.env.packageInfo.version,
+      config: coreSetup.legacy.pluginExtendedConfig,
+      kibanaConfig: (kibanaConfig as any) as KibanaConfig,
+      callCluster: retryCallCluster(adminClient.callAsInternalUser.bind(adminClient)),
+    });
+
+    return ({} as any) as Promise<SavedObjectsServiceSetup>;
+  }
+
+  public async start(core: SavedObjectsStartDeps): Promise<SavedObjectsServiceStart> {
+    this.logger.debug('Starting SavedObjects service');
+
+    /**
+     * Note: We want to ensure that migrations have completed before
+     * continuing with further Core startup steps that might use SavedObjects
+     * such as running the legacy server, legacy plugins and allowing incoming
+     * HTTP requests.
+     *
+     * However, our build system optimize step and some tests depend on the
+     * HTTP server running without an Elasticsearch server being available.
+     * So, when the `--skip-migrations` flag is true, we skip migrations all-
+     * together.
+     */
+    const cliArgs = this.coreContext.env.cliArgs;
+    const skipMigrations = cliArgs.optimize || cliArgs.skipMigrations;
+    await this.migrator!.awaitMigration(skipMigrations);
+
+    return { migrator: this.migrator as KibanaMigrator };
+  }
+
+  public async stop() {}
```

Comment:
`awaitMigration` takes some time. should stop terminate migration process?

Reply:
Stopping migrations while they are running will probably lead to data loss (worst case) or require manually deleting `.kibana_n` indexes to recover (best case). So in a way it would be good to wait for all migrations to finish before stopping. However I'm not sure how to handle this together with the es retry logic. If ES is down or goes down while migrations are running migrations will never finish.

I guess we could log a warning to the console like "Stopping saved objects service, waiting for migrations to complete". If someone really wants to stop Kibana because it seems unresponsive they can still do a `kill -9`. What do you think?

Reply:
Talked to Mike Cote who had a good suggestion than on `stop()` we cleanup any in-progress migration by pointing the `.kibana` alias to the previous known good `.kibana_n` index (we could rename the the half-migrated `kibana_n+1` index to indicate failure or delete it). I'll add it as a task for when we expose SavedObjectMigrations in NP. The behaviour in this PR is the same as the current behaviour in master.